src/codewhisperer/util/supplementalContext/rankBm25.ts

```diff
@@ -5,46 +5,126 @@
-    /** The document is originally scoreed. */
-    index: number
+export interface BM25Document {
+    content: string
     /** The score that the document recieves. */
     score: number
-}
 
-export function performBM25Scoring(corpusList: string[], query: string): BMDocument[] {
-    const wordFrequency: WordFrequency = initialize(corpusList)
-    const idf: IDF = calcIdf(wordFrequency)
-    const scoredDocs: BMDocument[] = getScore(wordFrequency, idf, query)
-    // Sort the result in descending order of scores.
-    return scoredDocs.sort((a: BMDocument, b: BMDocument) => b.score - a.score)
+    index: number
 }
 
-interface FrequencyMap {
-    [word: string]: number
-}
+export abstract class BM25 {
+    protected k1: number
+    protected b: number
+    protected epsilon: number
+
+    protected corpus: string[]
+    protected corpusSize: number
+    protected avgdl: number
+    protected idf: Map<string, number>
+    protected docLen: number[]
+    protected docFreqs: Map<string, number>[]
+    protected nd: Map<string, number>
+
+    constructor(
+        corpus: string[],
+        tokenizer: (str: string) => string[] = tokenize,
+        k1: number,
+        b: number,
+        epsilon: number
+    ) {
+        this.corpus = corpus
+        this.corpusSize = corpus.length
+        this.docLen = []
+        this.idf = new Map()
+        this.docFreqs = []
+        this.nd = new Map()
+        this.k1 = k1
+        this.b = b
+        this.epsilon = epsilon
+
+        let numDoc = 0
+        corpus
+            .map(document => {
+                return tokenizer(document)
+            })
+            .forEach(document => {
+                this.docLen.push(document.length)
+                numDoc += document.length
+
+                const frequencies = new Map<string, number>()
+                document.forEach(word => {
+                    frequencies.set(word, (frequencies.get(word) || 0) + 1)
+                })
+                this.docFreqs.push(frequencies)
+
+                frequencies.forEach((freq, word) => {
+                    this.nd.set(word, (this.nd.get(word) || 0) + 1)
+                })
+            })
+
+        this.avgdl = numDoc / this.corpusSize
+
+        this.calIdf(this.nd)
+    }
 
-interface WordFrequency {
-    docLen: number[]
-    docFreqs: FrequencyMap[]
-    corpusSize: number
-    avgDl: number
-    nd: FrequencyMap
-}
+    abstract calIdf(nd: Map<string, number>): void
 
-interface IDFMap {
-    [word: string]: number
-}
+    abstract score(query: string): BM25Document[]
 
-interface IDF {
-    idf: IDFMap
-    averageIdf: number
+    topN(query: string, n: number): BM25Document[] {
+        const notSorted = this.score(query)
+        const sorted = notSorted.sort((a, b) => b.score - a.score)
+        return sorted.slice(0, Math.min(n, sorted.length))
+    }
 }
 
-const BM25Configs = {
-    epsilon: 0.25,
-    k1: 1.5,
-    b: 0.75,
+export class BM25Okapi extends BM25 {
```

Comment:
Why the abstract base? Will there be many concrete implementations? There's _a lot_ of protected fields too, seems like the class only exists for initialization.

Reply:
yea, it's aligning with JB implementation and make it abstract because there are different variants of bm25 exist, I want it extendable so that if we decide to use different variants later it would be easier to just adding another subclass with its own calIdf.

Reply:
like there variants https://github.com/dorianbrown/rank_bm25/blob/master/rank_bm25.py

Reply:
is "many protected fields" an anti-pattern Jaden? I thought portected is fine since only subclass has access to it.

Reply:
basically what I did on JB and VSC change now is to do an exact same implementation as the original python library.

Reply:
> I thought portected is fine since only subclass has access to it.

I view having lots of `protected` state an anti-pattern because it becomes harder to reason about the entire state of an object when multiple things can change it. Protected immutable state doesn't have this issue.